File: procon/prime.MillerRabin.rs
/// Prime numbres - Miller Rabin Test
// @rand.lcg.rs
// @int.mod.pow.rs

// is_prime?
fn mrtest(n: u64) -> bool {
    if n < 2 {
        false
    } else if n < 4 {
        true
    } else if n % 2 == 0 {
        false
    } else {
        let mut rand = Rand(0);
        let mut d = n - 1;
        while d % 2 == 0 { d /= 2 }
        for _ in 0..100 {
            let a = rand.next() % (n - 1) + 1;
            let mut y = powmod(a, d, n);
            let mut t = d;
            while t != n - 1 && y != 1 && y != n - 1 {
                y = (y * y) % n;
                t <<= 1;
            }
            if y != n - 1 && t % 2 == 0 {
                return false;
            }
        }
        true
    }
}

File: procon/prime.MillerRabin.cc
namespace Prime {
  template<typename Int>
  Int powmod(Int a, Int k, Int m) {
    if (k == 0) return 1;
    Int t = powmod(a, k/2, m);
    t = (t * t) % m;
    if (k&1) t = (t * a) % m;
    return t;
  }

  // Miller-Rabin test
  template<typename Int=int>
  bool test(Int n) {
    if (n<2) return false;
    if (n<4) return true;
    if (!(n&1)) return false;

    random_device dev;
    mt19937 rand(dev());
    Int d = n-1;
    while (!(d&1)) d/=2;
    rep (_, 100) {
      Int a = rand()%(n-1) + 1;
      Int y = powmod<Int>(a,d,n);
      Int t = d;
      while (t!=n-1 and y!=1 and y!=n-1) {
        y = (y * y) % n;
        t <<= 1;
      }
      if (y!=n-1 and !(t&1)) return false;
    }
    return true;
  }
}
